sycl/include/CL/sycl/intel/group_algorithm.hpp

```diff
@@ -890,6 +900,39 @@ template <typename Group> bool leader(Group g) {
 #endif
 }
 
+template <typename Group> void barrier(Group, memory_scope scope) {
+  static_assert(sycl::detail::is_generic_group<Group>::value,
+                "Group algorithms only support the sycl::group and "
+                "intel::sub_group class.");
+#ifdef __SYCL_DEVICE_ONLY__
+  auto MemoryScope = sycl::detail::spirv::getScope(scope);
+  auto ExecutionScope = sycl::detail::spirv::group_scope<Group>::value;
+  __spirv_ControlBarrier(ExecutionScope, MemoryScope,
+                         __spv::MemorySemanticsMask::AcquireRelease |
+                             __spv::MemorySemanticsMask::SubgroupMemory |
+                             __spv::MemorySemanticsMask::WorkgroupMemory |
+                             __spv::MemorySemanticsMask::CrossWorkgroupMemory);
```

Comment:
I don't know whether this is important or not, but this barrier is different from both `work_group_barrier` and `sub_group_barrier` OpenCL functions.

I will probably list all parameters, even they are the same, just because representation in SPIR-V and OpenCL is a bit different and it is easy to make a mistake

Barrier enqueues control barrier, scope of which is defined by `Group` argument. In OpenCL this is defined by built-in itself (`work_group_barrier` vs `sub_group_barrier`), but functionality is preserved.

Barrier also enqueues memory fences, which has one or two (some overloads from OpenCL 2.0) parameters:
- mem_fence_flags: defines address spaces, which needs to be synchronized. Here it falls into `MemorySemanticsMask` of `ControlBarrier` SPIR-V instruction and user has no control over that. Please also note that both OpenCL and SPIR-V support synchronization of memory specific to images (`ImageMemory` in SPIR-V and `CLK_IMAGE_MEM_FENCE` in OpenCL)
- memory_scope: defines scope of memory (work-item, sub-group, work-group, whole device or all svm devices) which needs to be synchronized. Here it is named in the same way and can be configured by user as well

I've just realized that this might be done on purpose to align with extended atomics proposal, where we proposed single happens-before relationship for all address space:

> The fact that atomic operations in SYCL 1.2.1 obey separate happens-before relationships for global and local memory enables implementations to improve performance by limiting the scope of fences and other visibility operations (e.g. cache flushes) based on the address space to which an atomic operation is applied. The single happens-before relationship adopted by this proposal for consistency with C++ prevents this optimization, which may degrade performance on some architectures in some specific cases.

If so, would it be better to add similar clarification to `barrier` function description?

Reply:
Exactly, the design of `barrier` is supposed to be aligned with the single happens-before relationship for all address spaces.  I'll update the description as you suggest.

Did I get the order of the SPIR-V arguments correct?  The intended behavior is that it synchronizes the group, optionally issues a broader memory fence than the group, and always fences the full set of address spaces.

Reply:
I've tried to address this in 6a34aac3a0cc897cf7d36a803425da00534406a9.

Reply:
> Did I get the order of the SPIR-V arguments correct? The intended behavior is that it synchronizes the group, optionally issues a broader memory fence than the group, and always fences the full set of address spaces.

That seems to be correct

> optionally issues a broader memory fence than the group

However, I don't see how exactly this part is enforced in the code: my understanding that with this PR user should be able to set work-group barrier with sub-group fence within it

Reply:
> However, I don't see how exactly this part is enforced in the code: my understanding that with this PR user should be able to set work-group barrier with sub-group fence within it

Good point.  I was assuming that the device compiler would guarantee this, but I don't think that assumption is safe.  I've tried to address this in https://github.com/intel/llvm/pull/2198/commits/6f4424aca61444213ecf09cd4ef016d3716e3bdd.